### ISSUE TITLE
Add keyboard navigation

### ISSUE THREAD

**Issue by mbarlow12** [phetsims/molecules-and-light#143]
Like https://github.com/phetsims/ohms-law/issues/49, we would like keyboard navigation for the sim.

TODO:
- tool radio buttons
- conditionally remove 'Return Molecule' button from focus order when not displayed
- change focus order

**Comment by jessegreenberg** [comment]
@mbarlow12 I pushed our changes we made today in the above commit.

**Comment by mbarlow12** [comment]
@emily-phet How should the focus order appear here? It's pretty far out of order at the moment.

**Comment by jessegreenberg** [comment]
@emily-phet, @mbarlow12 and I were also wondering:

- What should happen to the focus on the "Return Molecule" button when the button disappears.  For now, @mbarlow12 placed focus on the photon emission slider when the button disappears.
- When the "step" button is disabled, should it be removed from the focus order?  For now, @mbarlow12 removed it from focus order when disabled for now.

We like the above behavior, but do you agree this makes sense?

**Comment by jessegreenberg** [comment]
6/27/17 meeting:

- We decided that radio button groups should work for the selection and navigation of photon emitters and molecules.
- We decided that the navigation order is good as it is now based on interaction patterns from interviews conducted by @arouinfar.
- Make sure that invisible/disabled buttons are still in the DOM (though not focusable) so screen reader users can find them even when they are not interactive.
- The focus highlight around the Play/Pause/step buttons should be a circle (like the ResetAllButton)
- For radio buttons with a visible label, focus highlight should go around the whole label.
- Make the focus highlight bigger for radio buttons in both groups so that it doesn't touch or overlap the "selected" highlight that currently exists.
- The focus highlight around the emission rate control slider should be larger (have more padding)
- Add a background in between the emission rate control slider and the focus highlight so that the focus highlight is easier to see.

**Comment by jessegreenberg** [comment]
![capture](https://user-images.githubusercontent.com/6396244/27600663-f924c572-5b29-11e7-81a0-2c4c653c8887.PNG)

@arouinfar @amanda-phet @amyh-phet @emily-phet regarding the last point, what should the background look like?  Should it be a grey rectangle that is under the focus highlight area?

**Comment by amanda-phet** [comment]
I think @emily-phet mentioned a white rectangle, but an opaque gray rectangle might help to de-clutter while also not being overly distracting.

**Comment by emily-phet** [comment]
@jessegreenberg I agree with Amanda that gray would be a good option to explore.

**Comment by jessegreenberg** [comment]
@mbarlow12 I fixed the Scenery problem with the above commit, thanks for catching that.

**Comment by mbarlow12** [comment]
Added the last two items in checklist. @emily-phet @arouinfar, the highlight seems a bit big with the above changes, but I'll defer to your judgment on that.

**Comment by emily-phet** [comment]
@mbarlow12 - Great! Can you share a link or screenshot with the new changes?

**Comment by mbarlow12** [comment]
@emily-phet here you go:

<img width="1236" alt="screen shot 2017-07-18 at 11 03 54 am" src="https://user-images.githubusercontent.com/5711147/28329917-df719de6-6ba8-11e7-92a0-91d1a1c33c37.png">

One last question, are we still implementing a custom fn & page up/down behavior? I saw some notes about this but haven't determined if it was discussed in detail and what that  behavior should be.

**Comment by emily-phet** [comment]
@mbarlow12 The screenshot looks great! The slider thumb with focus really stands out nicely. 

- Regarding custom fn and pg up/pg down behavior, it would be useful to be able to specify the step size when fn is used. Essentially, we'd like to be able to have a way to select a smaller step size for use with the shift modifier key, and a to select a larger step size for use with the fn modifier key.
For RIAW and Ohm's Law, the default fn behavior worked fine, but in further discussion it was brought up that PhET has a lot of sliders (like the one on the light source in this sim) that might have 10 or fewer steps total, and in that case the fn key with the default behavior may not work as expected (consistently giving larger step sizes). 

- Regarding pg up/pg down behavior, we want to keep that with the default behavior of jumping to the min/max of the slider range.

**Comment by jessegreenberg** [comment]
We could have an option called ```pageKeyboardStep``` that replaces the hard coded calculation in the keydown listener in HSlider.  We will also want getters and setters for ```pageKeyboardStep``` much like the ```setShiftKeyboardStep```and ```getShiftKeyboardStep``` that are already there.  @mbarlow12 let me know if you have questions while working on this.

**Comment by mbarlow12** [comment]
@amyh-phet Are you referring to the END/HOME behavior for jumping to the min/max values? @jessegreenberg and I don't think we can separate the 'fn' key specifically from PageUp/Down. But using HOME/END as the min/max jump, we can then customize the PageUp/Down steps at will.

**Comment by jessegreenberg** [comment]
@mbarlow12 added the ability to use custom step sizes for page up/page down.  I think the next steps are to 
- Test it out in MAL, making the slider in this sim behave such that
  - Holding shift moves the slider 1/20 of the slider range
  - Default step size is 1/10 of the slider range
  - page up/page down move 1/5 of the slider range
- Have @mbarlow12 test is account on spot by building a version of MAL with keyboard nav for @emily-phet @terracoda @arouinfar and @amanda-phet to review.

@mbarlow12 can you please change the slider to behave like the above with the keyboard and then we can try a dev build?

**Comment by mbarlow12** [comment]
Added the appropriate options for the Emission Slider for the changes requested in https://github.com/phetsims/molecules-and-light/issues/143#issuecomment-324113701.

**Comment by jessegreenberg** [comment]
Thanks @mbarlow12.  We tried to deploy to spot but found that the account didn't have the right permissions.  Once that is sorted we will try again.

**Comment by mbarlow12** [comment]
Sim deployed:
http://www.colorado.edu/physics/phet/dev/html/molecules-and-light/1.4.0-dev.1/molecules-and-light_en.html?accesibility

Let us know your thoughts!

**Comment by arouinfar** [comment]
@mbarlow12 is the above link correct? Tried to test it in Chrome and Firefox, and I couldn't get the focus to appear when tabbing.

**Comment by mbarlow12** [comment]
Whoops! I misspelled accessibility. The link works now. Thanks!

**Comment by arouinfar** [comment]
The keyboard navigation is feeling really smooth. I tested the behavior described in https://github.com/phetsims/molecules-and-light/issues/143#issuecomment-324113701, and everything looks great @mbarlow12!

**Comment by amanda-phet** [comment]
Looks good to me.

**Comment by terracoda** [comment]
Feels good to me, too. 
One question, what is the little "pause-play" supposed to do?

**Comment by jessegreenberg** [comment]
That is our "step" button.  It steps the simulation incrementally so that students can observe dynamic behavior in more fine detail by "stepping" through changes.

**Comment by terracoda** [comment]
@jessegreenberg, thanks for the clarification. In that case, I think I found a small issue.

Pressing and holding the Enter key plays continuous steps just like pressing and holding down the button with the mouse. However, pressing and holding the Space key does do the same thing. I have to release the Space key and then the items move a tiny bit and then stop.

I think Enter and Space keys should behave the same. They should both do the same thing as pressing and holding the button down with the mouse.

**Comment by jessegreenberg** [comment]
Thanks @terracoda, Ill create an issue for that.

**Comment by terracoda** [comment]
Just tested in Safari 10.0.3, and I see that the Arrow keys do not cycle through the list of molecules, or the list of lights like they do in Firefox. I think cycling through is the expected behaviour, but I will have to check. I'll create an issue.

**Comment by mbarlow12** [comment]
Tested in Safari 10.1.2, and the arrows are working properly. It looks like the 10.1 patch in March tackled some new JS and form input issues.

**Comment by terracoda** [comment]
Awesome news @mbarlow12.

**Comment by mbarlow12** [comment]
Correct: I was confused about the desired behavior. @jessegreenberg clarified—new Safari does NOT jump back to the first item. Apologies!

**Comment by terracoda** [comment]
Ok, still an issue to discuss for us, then ;-)

**Comment by jessegreenberg** [comment]
@terracoda would you please make a design doc for a11y for this sim?

**Comment by terracoda** [comment]
@mbarlow12 and @jessegreenberg, here's a link to [Molecules and Light: A11y Design Main](https://docs.google.com/document/d/1r9D3vq4QvmmXkQ5G6HexpybUv_vL6LrYFb3wgPfUdWg/edit#heading=h.yycebzf92xvc)

May still need to add some considerations in this issue.

**Comment by terracoda** [comment]
@amanda-phet, @arouinfar , @emily-phet, @mbarlow12 and @jessegreenberg, ok, I think I got all considerations in the design document.

Should this issue be closed?

**Comment by terracoda** [comment]
@amanda-phet, @arouinfar , @emily-phet, @mbarlow12 and @jessegreenberg,
Should these buttons be in a group?

Can interested folks comment in the design document?

**Comment by terracoda** [comment]
Sorry, I need to clarify last comment https://github.com/phetsims/molecules-and-light/issues/143#issuecomment-330883674, 

I meant to say, 

Should the Play, Pause, Step buttons be in a group?

And the grouping of these buttons can just take place in the DOM. It does not have to affect the current implementation of the tab navigation order, nor the use of the Tab key to focus the buttons.

**Comment by jessegreenberg** [comment]
Thanks @terracoda! yes I think this issue can be closed.  Any further changes can be discussed and tracked in new issues.

> Should the Play, Pause, Step buttons be in a group?

I added a comment in the doc. Closing!
